### PR TITLE
feat: Handle failed uploads (UPL-60)

### PR DESCRIPTION
**Pull request by markflorkowski**
### To do:
- [x] infra -- add endpoint for marking failed uploads
- [x] merge #195 
- [x] server uploads
- [x] test

### CODE REVIEW

**Review by juliusmarminge — COMMENTED** [review]
Server uploads?

File: packages/uploadthing/src/sdk/utils.ts [hunk 1]
```diff
@@ -96,6 +105,21 @@ export const uploadFilesInternal = async (
       });
 
       if (!s3res.ok) {
+        // tell uploadthing infra server that upload failed
+        await fetch(
+          createRequestUrl({
+            url: generateUploadThingURL("/"),
```
Comment (juliusmarminge) [inline packages/uploadthing/src/sdk/utils.ts:111]: probably want a pathname to this url no?

this will just create a url like `https://uploadthing.com/?actionType=failure&slug=__server-uploaded`, no?

File: packages/uploadthing/src/internal/handler.ts [hunk 2]
```diff
@@ -223,7 +223,7 @@ export const buildRequestHandler = <
       return { status: 200 };
     }
 
-    if (!actionType || actionType !== "upload") {
+    if (!actionType || ["upload", "failure"].includes(actionType)) {
```
Comment (juliusmarminge) [inline packages/uploadthing/src/internal/handler.ts:226]:
```suggestion
      if (!actionType || !["upload", "failure"].includes(actionType)) {
```

?

File: packages/uploadthing/src/sdk/utils.ts (hunk 1 above)
Comment (juliusmarminge) [reply packages/uploadthing/src/sdk/utils.ts:111]: or are you trying to first hit the users backend which will then hit uploadthings server? seems a bit unnecessary?

Comment (markflorkowski) [reply packages/uploadthing/src/sdk/utils.ts:111]: image

File: packages/uploadthing/src/client.ts [hunk 3]
```diff
@@ -183,9 +187,25 @@ export const DANGEROUS__uploadFiles = async <TRouter extends FileRouter>(
...
+          slug: String(opts.endpoint),
+          actionType: "failure",
+        }),
+        {
+          method: "POST",
+          body: JSON.stringify({
+            fileKey: fields.key,
+          }),
+        },
+      );
```
Comment (juliusmarminge) [inline packages/uploadthing/src/client.ts:203]: afaik this calls your server first, and then that will call uploadthing infra? is the plan to add an `onUploadError` hook in the file route config which will make this step necessary, or can we just send the fetch directly to uploadthing?

Comment (juliusmarminge) [reply packages/uploadthing/src/client.ts:203]: would it maybe also make sense to throw here instead, and then in an `.allSettled` check for errors and send a single request to the server, instead of 1 per file?

Comment (markflorkowski) [reply packages/uploadthing/src/client.ts:203]: Yeah, I am not sure why I went the route of going through your server first tbh...

Comment (markflorkowski) [reply packages/uploadthing/src/client.ts:203]: Right, I did this because we don't have the API key on client.

File: packages/uploadthing/src/internal/handler.ts [hunk 4]
```diff
@@ -206,15 +206,15 @@ export const buildRequestHandler = <
...
 
-    const reqBody = (await req.json()) as {
-      file: UploadedFile;
-      files: unknown;
-      metadata: Record<string, unknown>;
-      input?: Json;
-    };
-
-    if (uploadthingHook === "callback") {
+    if (uploadthingHook && uploadthingHook === "callback") {
```
Comment (juliusmarminge) [inline packages/uploadthing/src/internal/handler.ts:209]:
```suggestion
    if (uploadthingHook === "callback")
```

File: packages/uploadthing/src/internal/handler.ts [hunk 5]
```diff
@@ -223,7 +223,7 @@ export const buildRequestHandler = <
       return { status: 200 };
     }
 
-    if (!actionType || actionType !== "upload") {
+    if (!actionType || !["upload", "failure"].includes(actionType)) {
```
Comment (juliusmarminge) [inline packages/uploadthing/src/internal/handler.ts:226]: Could we perhaps move this array out, mark it as const and then infer the type from it and use that in `createRequestUrl`

Comment (juliusmarminge) [reply packages/uploadthing/src/internal/handler.ts:226]: Also the error below is outdated

File: packages/uploadthing/src/internal/handler.ts [hunk 6]
```diff
@@ -233,119 +233,169 @@ export const buildRequestHandler = <
     }
 
     try {
-      const { files, input: userInput } = reqBody as {
-        files: string[];
-        input: Json;
-      };
+      if (actionType === "upload") {
```
Comment (juliusmarminge) [inline packages/uploadthing/src/internal/handler.ts:236]: I'd prefer a `switch` over `else if` but not sure what we use elsewhere

File: packages/uploadthing/src/internal/handler.ts [hunk 7]
```diff
@@ -233,119 +233,169 @@ export const buildRequestHandler = <
...
+        } catch (error) {
+          console.error(error);
+          return new UploadThingError({
+            code: "BAD_REQUEST",
+            message: "Invalid input",
+            cause: error,
+          });
+        }
+
+        const metadata = await uploadable._def.middleware({
```
Comment (juliusmarminge) [inline packages/uploadthing/src/internal/handler.ts:257]: This can throw errors

File: packages/uploadthing/src/internal/handler.ts [hunk 8]
```diff
@@ -233,119 +233,169 @@ export const buildRequestHandler = <
...
+          error: new UploadThingError({
+            code: "UPLOAD_FAILED",
+            message: `Upload failed for ${fileKey}`,
+          }),
+          fileKey,
+        });
+
+        return { status: 200 };
+      } else {
+        return { message: "invalid actionType", status: 500 };
```
Comment (juliusmarminge) [inline packages/uploadthing/src/internal/handler.ts:397]:
```suggestion
        return new UploadThingError({ ... })
```

File: packages/uploadthing/src/internal/handler.ts (hunk 7 above)
Comment (markflorkowski) [reply packages/uploadthing/src/internal/handler.ts]: so it should be wrapped in a try-catch?

Comment (juliusmarminge) [reply packages/uploadthing/src/internal/handler.ts]: Like it was before, and then return the UploadThing error

Comment (markflorkowski) [reply packages/uploadthing/src/internal/handler.ts]: The whole switch is wrapped in a try catch that does this? 
image

Comment (juliusmarminge) [reply packages/uploadthing/src/internal/handler.ts]: Ah, mobile diff view is hard to parse 😅

That doesnt quite make sense though does it? There are more things that can fail in that big block which would cause a sub-optimal error message?

Comment (markflorkowski) [reply packages/uploadthing/src/internal/handler.ts]: Yeah... I should really go through and clean up this logic....

File: packages/uploadthing/src/client.ts [hunk 9]
```diff
@@ -33,9 +37,12 @@ function fetchWithProgress(
...
 
-const createRequestPermsUrl = (config: { url?: string; slug: string }) => {
-  const queryParams = `?actionType=upload&slug=${config.slug}`;
-
+const createRequestUrl = (config: {
+  url?: string;
+  slug: string;
+  actionType: ActionType;
+}) => {
+  const queryParams = `?actionType=${config.actionType}&slug=${config.slug}`;
```
Comment (juliusmarminge) [inline packages/uploadthing/src/client.ts:45]: If we're going nit I'd say use some `URL` or `URLSearchParams` here instead of a raw string but thats very nit

File: packages/uploadthing/src/internal/handler.ts [hunk 10]
```diff
@@ -223,135 +224,209 @@ export const buildRequestHandler = <
...
-    if (!actionType || actionType !== "upload") {
+    if (!actionType || !VALID_ACTION_TYPES.includes(actionType)) {
       // This would either be someone spamming or the AWS webhook
       return new UploadThingError({
         code: "BAD_REQUEST",
         cause: `Invalid action type ${actionType}`,
-        message: `Expected "upload" but got "${actionType}"`,
+        message: `Expected ${VALID_ACTION_TYPES.map((x) => `"${x}"`)
+          .join(", ")
+          .replace(/,(?!.*,)/, " or")} but got "${"a"}"`,
```
Comment (juliusmarminge) [inline packages/uploadthing/src/internal/handler.ts:234]: Damn that regex replace is hardcore :)

File: packages/uploadthing/src/internal/handler.ts [hunk 11]
```diff
@@ -223,135 +224,209 @@ export const buildRequestHandler = <
...
+        case "failure": {
+          const { fileKey } = (await req.json()) as {
+            fileKey: string;
+          };
+
+          // Tell uploadthing to mark the upload as failed
+          const uploadthingApiResponse = await fetch(
+            generateUploadThingURL("/api/failureCallback"),
+            {
+              method: "POST",
```
Comment (juliusmarminge) [inline packages/uploadthing/src/internal/handler.ts:373]: Formatting?

File: packages/uploadthing/src/internal/handler.ts [hunk 12]
```diff
@@ -223,135 +224,209 @@ export const buildRequestHandler = <
...
+              return new UploadThingError({
+                code: "URL_GENERATION_FAILED",
+                message: "Unable to get presigned urls",
+                cause,
+              });
+            }
+          }
+
+          // Run the onUploadError callback
+          uploadable._def.onUploadError({
```
Comment (juliusmarminge) [inline packages/uploadthing/src/internal/handler.ts:406]: Have we documented if this can throw or not? Might be good to tell people what this function should do. If this throws and isnt wrapped in try-catch errors might be weird?

**Review by juliusmarminge — COMMENTED** [review]
I'll review properly when i get back on Saturday but looks good!

File: packages/uploadthing/src/internal/handler.ts (hunk 10 above)
Comment (markflorkowski) [reply packages/uploadthing/src/internal/handler.ts:234]: Every other way I tried of doing it seemed like too much code hahah

File: packages/uploadthing/src/internal/handler.ts (hunk 11 above)
Comment (markflorkowski) [reply packages/uploadthing/src/internal/handler.ts:373]: ?

File: packages/uploadthing/src/client.ts [hunk 13]
```diff
@@ -90,9 +103,10 @@ export const DANGEROUS__uploadFiles = async <TRouter extends FileRouter>(
 ) => {
   // Get presigned URL for S3 upload
   const s3ConnectionRes = await fetch(
-    createRequestPermsUrl({
+    createRequestUrl({
```
Comment (t3dotgg) [inline packages/uploadthing/src/client.ts:106]: nit: might be worth naming this "createS3RequestUrl" or something to make it clear what it's calling specifically

File: packages/uploadthing/src/internal/handler.ts [hunk 14]
```diff
@@ -223,135 +224,222 @@ export const buildRequestHandler = <
...
+                "Content-Type": "application/json",
+                "x-uploadthing-api-key": preferredOrEnvSecret,
+                "x-uploadthing-version": UPLOADTHING_VERSION,
+              },
+            },
+          );
+
+          if (!uploadthingApiResponse.ok) {
+            console.error("[UT] failed to mark upload as failed");
+            try {
```
Comment (t3dotgg) [inline packages/uploadthing/src/internal/handler.ts:387]: We do this a few times, would it be worth making a `safeParseJson` helper?

**Review by t3dotgg — APPROVED** [review]

File: packages/uploadthing/src/internal/handler.ts (hunk 11 above)
Comment (juliusmarminge) [reply packages/uploadthing/src/internal/handler.ts:373]: never mind, looked like indentation was incorrect on mobile 😅

File: packages/uploadthing/src/internal/handler.ts [hunk 15]
```diff
@@ -223,135 +224,222 @@ export const buildRequestHandler = <
...
         code: "BAD_REQUEST",
         cause: `Invalid action type ${actionType}`,
-        message: `Expected "upload" but got "${actionType}"`,
+        message: `Expected ${VALID_ACTION_TYPES.map((x) => `"${x}"`)
+          .join(", ")
+          .replace(/,(?!.*,)/, " or")} but got "${"a"}"`,
       });
     }
 
     try {
```
Comment (juliusmarminge) [inline packages/uploadthing/src/internal/handler.ts:238]: i don't quite like this outer try-catch block. what is it that can throw that we're not catching explicitely?

File: packages/uploadthing/src/internal/handler.ts [hunk 16]
```diff
@@ -223,135 +224,222 @@ export const buildRequestHandler = <
...
+            return new UploadThingError({
+              code: "INTERNAL_SERVER_ERROR",
+              message: "Failed to run onUploadError callback",
+              cause: error,
+            });
+          }
+
+          return { status: 200 };
+        }
+        default: {
```
Comment (juliusmarminge) [inline packages/uploadthing/src/internal/handler.ts:429]: should make sure we get typeerror if we forget handling a case in the future

```suggestion
        }
        actionType satisfies never;
        default: {
```

File: packages/uploadthing/src/internal/types.ts [hunk 17]
```diff
@@ -47,6 +47,7 @@ export interface AnyParams {
   _metadata: any; // imaginary field used to bind metadata return type to an Upload resolver
   _runtime: any;
   _errorShape: any;
+  _error: any; // used for onUploadError
```
Comment (juliusmarminge) [inline packages/uploadthing/src/internal/types.ts:50]: maybe `_errorFn` to properly differentiate it from `_errorShape`?

File: pnpm-lock.yaml [hunk 18]
```diff
@@ -1,4 +1,4 @@
-lockfileVersion: '6.0'
+lockfileVersion: '6.1'
```
Comment (juliusmarminge) [inline pnpm-lock.yaml:1]: 👀

**Review by juliusmarminge — APPROVED** [review]

**Review by t3dotgg — APPROVED** [review]

**Comment by markflorkowski** [comment]
image

**Review by t3dotgg — APPROVED** [review]
